libraries/AP_AIS/AP_AIS.h

```diff
@@ -0,0 +1,138 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <AP_Param/AP_Param.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <AP_Common/AP_ExpandingArray.h>
+
+#ifndef HAL_AIS_ENABLED
+#define HAL_AIS_ENABLED !HAL_MINIMIZE_FEATURES
+#endif
+
+#if HAL_AIS_ENABLED
+
+#define AIVDM_BUFFER_SIZE 10
+#define AIVDM_PAYLOAD_SIZE 65
+
+class AP_AIS
+{
+public:
+    AP_AIS();
+
+    /* Do not allow copies */
+    AP_AIS(const AP_AIS &other) = delete;
+    AP_AIS &operator=(const AP_AIS&) = delete;
+
+    // return true if AIS is enabled
+    bool enabled() const { return AISType(_type.get()) != AISType::NONE; }
+
+    // Initialize the AIS object and prepare it for use
+    void init();
+
+    // update AIS, expected to be called at 20hz
+    void update();
+
+    // send mavlink AIS message
+    void send(mavlink_channel_t chan);
+
+    // parameter block
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+
+    // parameters
+    AP_Int8 _type;             // type of AIS receiver
+    AP_Int16 _max_list;        // maximum number of vessels to track at once
+    AP_Int16 _time_out;        // time in seconds that a vessel will be dropped from the list
+    AP_Int16 _options;         // options bitmask
```

Comment:
as mentioned on discord, I'd probably name this param "AIS_DEBUG_LVL" instead of options but it's not a blocker.

Reply:
changed to `AIS_LOGGING`